Also record the raw device events through BPF

bentiss opened the pull request:
There are multiple case where the events forwarded to hidraw are not the pristine ones from the device:
- if a HID-BPF filter is inserted and changes the incoming data
- if a driver changes the data in its `.raw_event()`

Let's inject a new HID-BPF filter in front of anyone else, and use it to retrieve the data from the device.

This is controlled by a new `--bpf` flag or will also be enabl;ed if we detect that a HID-BPF filter is in used on the given device.

Further improvements can be done:
- ~support more than 64 bytes of data by injecting more than one element into the ringbuf~ done
- ~detect cases where the chances of the event stream not to be tempered by the kernel (no HID-BPF filters and hid-generic in use or hid-multitouch)~

whot commented on src/main.rs:
```diff
@@ -1143,12 +1228,41 @@ fn hid_recorder() -> Result<()> {
+    let maps = skel.maps();
+
+    // We need to keep _link around or the program gets immediately removed
+    let _link = maps.hid_record().attach_struct_ops()?;
```
even if it's not behind a feature I reckon we should still have it behind a commandline argument to avoid confusing users with output that's near-always identical?

Plus right now this would fail if hid-bpf isn't enabled, right?

Possibly other idea: warn or only start printing bpf event data if the data differs from the kernel data?

bentiss replied:
After our meeting this morning: we should only enable the BPF output if there is a bpf loaded attached to the device. This way things will stand out much more when we see the `B:` line.

But adding a command line argument is also interesting because we will be able to request it when we suspect the kernel driver is acting behnd our back.

bentiss replied:
done in the latest push

whot reviewed (commented):
still haven't managed to look at the first patch but here's a few review comments anyway :)

bentiss commented on Cargo.toml:
```diff
@@ -20,3 +20,7 @@ chrono = "0.4.38"
 hidreport = "0.4.0"
 hut = "0.2.0"
 human-sort = "0.2.2"
+libbpf-rs = "0.23"
```
I just checked, it doesn't require locally libbpf :)

However i requires the following packages to be able to compile under a brand new fedora 40 container:
```
dnf install -y cargo git-core elfutils-libelf-devel clang
```

So libbpf is not required but clang and libelf are...

And then:
```
$ ldd target/debug/hid-recorder   
	linux-vdso.so.1 (0x00007f0cdb4cb000)
	libelf.so.1 => /lib64/libelf.so.1 (0x00007f0cdb488000)
	libz.so.1 => /lib64/libz.so.1 (0x00007f0cdb46e000)
	libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007f0cdb44a000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f0cdae1e000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f0cdb4cd000)
	libzstd.so.1 => /lib64/libzstd.so.1 (0x00007f0cdad5f000)
```

So libbpf is statically compiled.

As for running behind a feature... not sure. I would rather keep this unconditionally, and if people complain about the compile time or requirements we can always provide binaries for the program that people can just pull.

whot commented on src/main.rs:
```diff
@@ -64,17 +157,24 @@ impl Styles {
 const MAX_USAGES_DISPLAYED: usize = 5;
 
 // Usage: cprintln!(Sytles::Data, <normal println args>)
```
```suggestion
// Usage: cprintln!(Styles::Data, <normal println args>)
```
might as well fix it while we're here

whot commented on src/main.rs:
```diff
@@ -64,17 +157,24 @@ impl Styles {
     }};
 }
 
+// Usage: cprint!(Sytles::Data, <normal println args>)
```
```suggestion
// Usage: cprint!(Styles::Data, <normal println args>)
```

whot commented on src/main.rs:
```diff
@@ -19,6 +21,75 @@ use hidreport::hid::{
+            return Ok(());
+        }
+
+        std::env::set_var("NO_COLOR", "1");
```
unrelated note because it's an existing bug: this does not handle `--color=always` together with an output file. meh, I guess but still...

bentiss replied:
filed #28 so we don't lose it

whot commented on src/main.rs:
```diff
@@ -1084,20 +1113,42 @@ fn read_events(path: &Path, rdesc: &ReportDescriptor) -> Result<()> {
+    let mut builder = libbpf_rs::RingBufferBuilder::new();
+    builder
+        .add(map_ringbuf, move |data| {
+            event_handler(data, &mut start_time)
```
weird, I'm surprised this works. You're passing a mutable ref to `start_time` into the builder but later in the `f.read()` arms we change `start_time`. Really surprised rust doesn't throw the borrow checker at us here.

bentiss replied:
Not sure how this is possible, but it turns out that the value we get here is not shared with the main thread. So if I update in hidraw event, the value is still None in the bpf handler. And the other way around.

I might as well get a different value.

whot commented on src/main.rs:
```diff
@@ -199,6 +199,10 @@ struct Cli {
     only_describe: bool,
 
+    /// Also grab the events from the device through HID-BPF.
+    #[arg(long, default_value_t = false)]
```
shouldn't this be a `true/false/auto`? Could re-use `clap::ColorChoice` here which allows for `always/never/auto`

whot commented on src/main.rs:
```diff
@@ -1029,33 +1124,45 @@ fn read_events(stream: &mut impl Write, path: &Path, rdesc: &ReportDescriptor) -
     let mut start_time: Option<Instant> = None;
     let mut last_timestamp: Option<Instant> = None;
     let mut data = [0; 1024];
+    let mut ringbuf: Option<libbpf_rs::RingBuffer> = None;
+
+    if let Some(map_ringbuf) = map_ringbuf {
+        let mut builder = libbpf_rs::RingBufferBuilder::new();
+        builder
+            .add(map_ringbuf, move |data| {
+                event_handler(data, &mut start_time)
+            })
+            .unwrap();
+        ringbuf = Some(builder.build().unwrap());
+    }
```
```suggestion
    let ringbuf: Option<libbpf_rs::RingBuffer> = if let Some(map_ringbuf) = map_ringbuf {
        let mut builder = libbpf_rs::RingBufferBuilder::new();
        builder
            .add(map_ringbuf, move |data| {
                event_handler(data, &mut start_time)
            })
            .unwrap();
        Some(builder.build().unwrap())
    } else {
        None
    }
```

though I think what should also work is:
```rs
let ringbuf = map_ringbuf.map(|map_ringbuf| {
   let mut builder ...
});
```

bentiss replied:
that last suggestion makes the code much cleaner, thanks!

whot commented on src/main.rs:
```diff
@@ -1255,19 +1263,33 @@ fn hid_recorder() -> Result<()> {
     }
 
-    let skel = open_skel.load().unwrap();
+    Ok(Some(open_skel.load().unwrap()))
```
can this fail? Do we need to drop the `unwrap()`?
```rs
Ok(Some(open_skel.load()?))
```

bentiss replied:
it can fail on some occasions, but if it fails, we should bail out (now that we got proper `--bpf` argument handling)

whot commented on src/main.rs:
```diff
@@ -1255,19 +1263,33 @@ fn hid_recorder() -> Result<()> {
     let rdesc = parse(&rdesc_file, &opts)?;
     if !cli.only_describe && path.starts_with("/dev") {
-        read_events(&path, &rdesc, maps.events())?
+        match preload_bpf_tracer(cli.bpf, &path)? {
```
if there's an error during the BPF loading (which AIUI can happen legitimately if HID-BPF is disabled) then we just bail out here instead of handling the `Result` and continuing without BPF.

bentiss replied:
If we try to load the bpf, it's either because there is already a bpf loaded, or if the user explicitely request the loading of the bpf. In both cases it seems appropriate to bail out and not continue.

whot commented on src/main.rs:
```diff
@@ -1093,23 +1200,87 @@ fn find_device() -> Result<PathBuf> {
+
+fn hid_recorder() -> Result<()> {
+    let cli = Cli::parse();
+
```
```suggestion
```

nitpick: too many empty lines

whot commented on src/main.rs:
```diff
@@ -1093,23 +1200,87 @@ fn find_device() -> Result<PathBuf> {
+        .replace([':', '.'], "_");
+
+    let bpffs = PathBuf::from("/sys/fs/bpf/hid/").join(sysfs_name);
+
```
```suggestion
```
nitpick: too many empty lines

whot commented on src/main.rs:
```diff
@@ -1093,23 +1200,87 @@ fn find_device() -> Result<PathBuf> {
+        *start_time
     };
 
+    let elapsed = start_time.unwrap().elapsed();
+
+    if event.length == 0 {
+        return 1;
+    }
```
```suggestion
    if event.length == 0 {
        return 1;
    }
    
    let elapsed = start_time.unwrap().elapsed();
```

we don't need elapsed if event.length is 0.


Arguably - do we want to set `*start_time` if `event.length == 0`? seems like a niche case and it's cleaner to do the various event checks first before getting into the time/elapsed/etc. handling

whot reviewed (commented):
This is mostly ready now, with a few more nitpicks and minor changes please, thanks.

bentiss commented:
And I've now fixed the github workflow.

So I think we can have another look at this now.

Note: I've also added a commit to support packets longer than 64 bytes, but the syntax is not very clean as there will be 2 `B:` tags: one for the beginning with the timestamp and the total size, and all the rest will have just `B:` and the remaining bytes.

whot commented on .github/workflows/main.yml:
```diff
@@ -8,6 +8,11 @@ jobs:
+        run: |
+          sudo apt-get update
+          sudo apt-get install llvm libelf-dev clang
+          clang --version
```
i'm going to say this is going to be an issue for us at some point - most of our users won't have llvm/clang installed so they'll all run into cryptic build errors. We need some user-friendly "Warning, you need these packages installed" during the cargo build process.

And/or a github artifact for direct download (filed as issue #29 so we don't forget)

bentiss replied:
One solution could be what is done in the kernel for pre-loaded bpf: we generate the skeleton and then check in that output. Then we add a flag to cargo to use the embedded generated skeleton or regenerate it (with potentially some way to detect if they are out of sync, maybe with `git diff`).

> And/or a github artifact for direct download (filed as issue https://github.com/hidutils/hid-recorder/issues/29 so we don't forget)

Thanks for that!

bentiss replied:
I've managed to turn `clang` optional by adding the generated `.skel.rs` to the source tree and calling or not `SkeletonBuilder`.

However, I can not make `libelf-dev` (or `elfutils-libelf-devel` on fedora) optional: when libppf-rs is pulled, it pulls libbpf-sys which compiles itself and fails at `libelf.h` not found.

 I could hide the bpf functionality behind a feature, which would make `libbpf-rs` optional, but then we would need to figure out whether we want it enabled by default or not. And in the end, we probably just ask users to download the binary from #29, so not sure it's worth the effort.

bentiss commented on src/bpf/hidrecord.bpf.c:
```diff
@@ -1,46 +1,14 @@
 #define BPF_F_BEFORE (1U << 3)
 
-SEC("struct_ops/hid_device_event")
+SEC("?struct_ops/hid_device_event")
```
It's a leftover from when I tried to have only one bpf program instead of two.
You can add a `?` in the SEC which tells libbpf to not load the program (autoload is `false`). However it wasn't working as expected for 2 reasons:
- you can not not autoload the struct_ops declaration (technically you can manually do that later in the code)
- it complained that the `syscall` to attach the tracing version of the filter was not having the proper kernel kfunc.
So I went the easy road and split the file in 2.

But the more I think of it, the more I think I must have done something wrong and might still be able to have a single `.bpf.c` file.

bentiss commented on src/main.rs:
```diff
@@ -1304,13 +1412,24 @@ fn hid_recorder() -> Result<()> {
-            None => read_events(&path, &rdesc, None)?,
-            Some(skel) => {
+            HidBpfSkel::None => read_events(&path, &rdesc, None)?,
+            HidBpfSkel::StructOps(skel) => {
```
tracing is discontinued in v6.11. So struct_opt is mandatory. The question is more do we want to keep tracing, but given that the vas majority of users is using tracing, we should do both.

bentiss commented on src/bpf/hidrecord.h:
```diff
@@ -23,20 +27,31 @@ extern __u8 *hid_bpf_get_data(struct hid_bpf_ctx *ctx,
 
-	event->length = hctx->size;
-	__builtin_memcpy(event->data, data, sizeof(event->data));
+	for (i = 0; i * PACKET_SIZE < hctx->size && i < 256; i++) {
```
IIRC that's a value for the verifier to have an upper bound. If we use packet_size, it's an unknown at compile time and so the verifier assumes `INT_MAX` and complains about too many loops.

I'll try to use a better `#define` for this.

bentiss commented on src/main.rs:
```diff
@@ -1278,17 +1282,26 @@ fn event_handler(data: &[u8]) -> ::std::os::raw::c_int {
+            bytes,
+        );
+    } else {
+        cprintln!(Styles::Bpf, "B:                  {}", bytes,);
```
> We have the packet_number in the struct, why don't we collect all packets and print them all as one line?

that would make the producer harder. We are in a callback, so we need some sort of static buffer, and these are notably hard to do in rust. I'll see if I can pull it from my hat.

> Make the number count down so it indicates remaining packets and then we know when to stop and print the \n.

So you mean use `cprint!()`, not `cprintln!()`? What if we get a partial bpf event and then move to hidraw?

Using the remaining packets can be already computed because I have the packet count and the packet number in the event, so it's just an difference away.

bentiss replied:
So I could just use a:
```
let mut bpf_vec = Vec::new();
...
        builder.add(map_ringbuf, |data| event_handler(data, &mut bpf_vec)).unwrap();
```

And in the `event_handler()` call, I have this shared buffer available :)

So no interaction between hidraw and BPF, and we are good

whot commented on src/main.rs:
```diff
@@ -1139,6 +1186,48 @@ fn find_device() -> Result<PathBuf> {
 
+fn event_handler(data: &[u8]) -> ::std::os::raw::c_int {
+    static START_TIME_ONCE: Once = Once::new();
+    static mut START_TIME: Option<Instant> = None;
```
I'm wondering - have we simply reached the point where we need a `struct HidRecorder` which contains the `start_time` and we pass that into the event handler and `parse_input_report()` (so we share the same start time instead of initializing it twice). That would make life a lot easier...

That struct can also contain whatever other values we are forcibly trying to make `static` with a lot of effort.

whot replied:
ended up doing this with a `OnceCell` instead, see #34

whot commented on src/main.rs:
```diff
@@ -1139,6 +1253,166 @@ fn find_device() -> Result<PathBuf> {
+    /* we strip out the 3 following lines that happen when the kernel
+     * doesn't support HID-BPF struct_ops
+     */
+    if msg.contains("struct bpf_struct_ops_hid_bpf_ops is not found in kernel BTF")
+        || msg.contains("failed to load object 'hidrecord_bpf'")
+        || msg.contains("failed to load BPF skeleton 'hidrecord_bpf': -2")
+    {
+        return;
+    }
```
```rust
    let ignore_msgs = vec!["struct bpf  ..."];
    if ignore_msg.iter().any(|ignore| msg.contains(ignore)) {
        return;
    }
```
would be nicer

whot replied:
done in #34, please feel free to cherry-pick and squash

whot commented on src/main.rs:
```diff
@@ -1088,20 +1117,38 @@ fn read_events(path: &Path, rdesc: &ReportDescriptor) -> Result<()> {
+                        if start_time.is_none() {
+                            start_time = last_timestamp;
+                        }
+                        let _ = ringbuf.consume();
```
honestly, this is a big ugly. We're expecting an error from the fd read and then fall back to the bpf fd. The correct approach for `poll()` would be to check which one of the `pollfds` has `PollFlags::revents()` set to `PollFlags::POLLIN`. So something like:

```rust
if poll(&mut pollfds, timeout)? > 0 {
   if pollfds[0].revents().map_or(false, |f| f.intersects(PollFlags::POLLIN)) {
       match f.read() {
       }
   }
   // same for pollfds[1]
}
```

(I typed this here so it probably won't compile like this but you get the gist)

whot replied:
done in #34, please feel free to cherry-pick and squash

whot commented on src/bpf/hidrecord.h:
```diff
@@ -23,20 +29,31 @@ extern __u8 *hid_bpf_get_data(struct hid_bpf_ctx *ctx,
 
-	event = bpf_ringbuf_reserve(&events, sizeof(*event), 0);
-	if (!event)
+	if (hctx->size < 0 || packet_count > 255)
```
`packet_count >= HID_MAX_PACKET`?

whot commented on build.rs:
```diff
@@ -1,18 +1,41 @@
+        );
+        println!("cargo:warning= changes in 'src/bpf/*' will not be taken into account");
+        return;
+    }
```
*handwavy*: instead of reimplementing `which` wouldn't it better to try to compile and detect a compilation failure? Not sure the error gives us enough detail for this thoug...

whot replied:
done in #34, please feel free to cherry-pick and squash

whot commented on src/bpf/hidrecord.skel.rs:
```diff
@@ -0,0 +1,693 @@
+
+    const DATA: &[u8] = &[
+        127, 69, 76, 70, 2, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 247, 0, 1, 0, 0, 0, 0, 0, 0, 0,
+        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 40, 28, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 64, 0, 0, 0, 0,
```
I'm getting a lot of diff in these files as soon as I run cargo build. It's just bytes but also means I need to be careful with git commit, git diff is messy, etc.

Haven't tracked down why yet but it's probably a version mismatch but either way it's annoying. git doesn't seem to have an option for "track this file but don't change it".

whot replied:
done in #34, please feel free to cherry-pick and squash

whot commented:
Let's get this merged and fix up any remaining things later, thanks for the patches!
